Sources/NIOHTTP1/HTTPServerPipelineHandler.swift

```diff
@@ -0,0 +1,227 @@
+/// half-close. Having this `ChannelHandler` be aware of TCP half-close makes it easier
+/// to build HTTP servers that are resilient to this kind of behaviour.
+///
+/// The TCP half-close handling is done by buffering the half-close notification along
+/// with the HTTP request parts. The half-close notification will be delivered in order
+/// with the rest of the reads. If the half-close occurs either before a request is received
+/// or during a request body upload, it will be delivered immediately. If a half-close is
+/// received immediately after `HTTPServerRequestPart.end`, it will also be passed along
+/// immediately, allowing this signal to be seen by the HTTP server as early as possible.
+public final class HTTPServerPipelineHandler: ChannelDuplexHandler {
+    public typealias InboundIn = HTTPServerRequestPart
+    public typealias InboundOut = HTTPServerRequestPart
+    public typealias OutboundIn = HTTPServerResponsePart
+    public typealias OutboundOut = HTTPServerResponsePart
+
+    public init() { }
+
+    /// The state of the HTTP connection.
+    private enum ConnectionState {
+        /// We are waiting for a HTTP response to complete before we
+        /// let the next request in.
+        case responseEndPending
+
+        /// Neither a request nor a response are complete. This can be either
+        /// because nothing is happening on the connection, or because the request
+        /// and response are simultaneously progressing and have not completed.
+        case idle
+
+        /// The server has responded early, before the request has completed. We need
+        /// to wait for the request to complete, but won't block anything.
+        case requestEndPending
+
+        mutating func responseEndReceived() {
+            switch self {
+            case .responseEndPending:
+                // Got the response we were waiting for.
+                self = .idle
+            case .idle:
+                // We got a response while still receiving a request, which we have to
+                // wait for.
+                self = .requestEndPending
+            case .requestEndPending:
+                preconditionFailure("Received second response")
+            }
+        }
+
+        mutating func requestEndReceived() {
+            switch self {
+            case .requestEndPending:
+                // Got the request end we were waiting for.
+                self = .idle
+            case .idle:
+                // We got a request and the response isn't done, wait for the
+                // response.
+                self = .responseEndPending
+            case .responseEndPending:
+                preconditionFailure("Received second request")
+            }
+        }
+    }
+
+    /// The events that this handler buffers while waiting for the server to
+    /// generate a response.
+    private enum BufferedEvent {
+        /// A channelRead event.
+        case read(NIOAny)
```

Comment:
maybe call it `channelRead` to ensure this is not confused with `read` method in `ChannelOutboundHandler`. Just a suggestion tho

Reply:
@Lukasa did you see this and decide that its ok as it is ? Thats fine, just wanted to ensure you not missed it

Reply:
Oh lol, I only did it for the *other* enum. Fixed.